Intermittent build failures with  "error CS0104: `X` is an ambiguous reference between `System.X` and `System.X`"

**Version Used**: 
- MSBuild version = "17.1.0+ae57d105c"
- dotnet sdk 6.0.200

**Steps to Reproduce**:

1. Clone https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6
2. Repro steps in the repo README _(TL;DR: do a sonar analysis when running the build, which injects sonar Roslyn analyzers; run the build multiple times until it fails; it intermittently fails - only ~20% of the runs)_

**Expected Behavior**: build & analysis is successfull

**Actual Behavior**:
Build fails intermittently with "ambiguous reference" errors e.g. 

>(CoreCompile target) -> d:\support\S4MSB\AmbiguousTypesNet6\Class2.cs(13,32): error CS0104: 'TimeSpan' is an ambiguous reference between 'System.TimeSpan' and 'System.TimeSpan' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(14,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj] d:\support\S4MSB\AmbiguousTypesNet6\Entities.cs(19,16): error CS0104: 'DateTime' is an ambiguous reference between 'System.DateTime' and 'System.DateTime' [d:\support\S4MSB\AmbiguousTypesNet6\AmbiguousTypesNet6.csproj]

**Binary Logs**:

You can check the binary logs for a successful build and a failed one: https://github.com/duncanp-sonar/BugRepro_AmbiguousTypesNet6/tree/master/logs. 

Note: the `csc.exe` invocations are identical.

**Details**:

Initially reported on [sonar community](https://community.sonarsource.com/t/timespan-is-an-ambiguous-reference-between-system-timespan-and-system-timespan/59245/9).  Some users have reported other numbers for their projects (only 1 / 10 of the builds succeed). A user reported that by removing `ImplicitUsings`, the problem would stop manifesting itself.

Separately, this has been reported [on StackOverflow as well](https://stackoverflow.com/questions/71171933/error-cs0104-guid-is-an-ambiguous-reference-between-system-guid-and-system) (seems not to be related to the sonar Roslyn analyzers).

Can this be related to https://github.com/dotnet/roslyn/issues/59564 ?

Could you please help:
- Are you aware of non-deterministic behavior inside the compiler that could lead to the stochastic errors we are witnessing?
- Do you have any hints on what to investigate in order to drill down to the root cause of the issue? 

Thank you!

--- a/src/Compilers/Core/CodeAnalysisTest/CachingLookupTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/CachingLookupTests.cs
@@ -10,6 +10,7 @@ using System.Collections.Immutable;
 using System.Linq;
 using Microsoft.CodeAnalysis.Collections;
 using Microsoft.CodeAnalysis.Text;
+using Roslyn.Test.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.UnitTests
@@ -286,6 +287,84 @@ namespace Microsoft.CodeAnalysis.UnitTests
             retval = look1[key1];
             retval = look1[key2];
             retval = look1[key3];
+        }
+
+        [Fact]
+        [WorkItem(60259, "https://github.com/dotnet/roslyn/issues/60259")]
+        public void FullyPopulateRace_01()
+        {
+            CachingDictionary<int, FullyPopulateRaceHelper> lookup = null;
+            FullyPopulateRaceHelper item = null;
+            int itemAccessCount = 0;
+            lookup = new CachingDictionary<int, FullyPopulateRaceHelper>(getElementsOfKey, getKeys, EqualityComparer<int>.Default);
+
+            _ = lookup.Count;
+            Assert.NotNull(item);
+            Assert.Same(item, getItem());
+
+            ImmutableArray<FullyPopulateRaceHelper> getElementsOfKey(int id)
+            {
+                itemAccessCount++;
+                if (item is null && itemAccessCount == 1)
+                {
+                    item = getItem();
+                }
+
+                return ImmutableArray.Create<FullyPopulateRaceHelper>(new FullyPopulateRaceHelper());
+            }
+
+            HashSet<int> getKeys(IEqualityComparer<int> comparer)
+            {
+                return new HashSet<int>(new[] { 1 }, comparer);
+            }
+
+            FullyPopulateRaceHelper getItem()
+            {
+                return lookup[1][0];
+            }
+        }
+
+        private class FullyPopulateRaceHelper { }
+
+        [Fact]
+        [WorkItem(60259, "https://github.com/dotnet/roslyn/issues/60259")]
+        public void FullyPopulateRace_02()
+        {
+            CachingDictionary<int, FullyPopulateRaceHelper> lookup = null;
+            FullyPopulateRaceHelper item = null;
+            int itemAccessCount = 0;
+            lookup = new CachingDictionary<int, FullyPopulateRaceHelper>(getElementsOfKey, getKeys, EqualityComparer<int>.Default);
+
+            Assert.Empty(lookup[2]);
+            _ = lookup.Count;
+            Assert.NotNull(item);
+            Assert.Same(item, getItem());
+
+            ImmutableArray<FullyPopulateRaceHelper> getElementsOfKey(int id)
+            {
+                if (id != 1)
+                {
+                    return ImmutableArray<FullyPopulateRaceHelper>.Empty;
+                }
+
+                itemAccessCount++;
+                if (item is null && itemAccessCount == 1)
+                {
+                    item = getItem();
+                }
+
+                return ImmutableArray.Create<FullyPopulateRaceHelper>(new FullyPopulateRaceHelper());
+            }
+
+            HashSet<int> getKeys(IEqualityComparer<int> comparer)
+            {
+                return new HashSet<int>(new[] { 1 }, comparer);
+            }
+
+            FullyPopulateRaceHelper getItem()
+            {
+                return lookup[1][0];
+            }
         }
     }
 }

--- a/src/Compilers/Core/Portable/Collections/CachingDictionary.cs
+++ b/src/Compilers/Core/Portable/Collections/CachingDictionary.cs
@@ -211,7 +211,7 @@ namespace Microsoft.CodeAnalysis.Collections
         /// </summary>
         /// <param name="existingMap">The existing map which may be null or a ConcurrentDictionary.</param>
         /// <returns></returns>
-        private IDictionary<TKey, ImmutableArray<TElement>> CreateFullyPopulatedMap(IDictionary<TKey, ImmutableArray<TElement>>? existingMap)
+        private IDictionary<TKey, ImmutableArray<TElement>> CreateFullyPopulatedMap(ConcurrentDictionary<TKey, ImmutableArray<TElement>>? existingMap)
         {
             Debug.Assert(IsNotFullyPopulatedMap(existingMap));
 
@@ -237,7 +237,7 @@ namespace Microsoft.CodeAnalysis.Collections
 
                     if (!existingMap.TryGetValue(key, out elements))
                     {
-                        elements = _getElementsOfKey(key);
+                        elements = existingMap.GetOrAdd(key, _getElementsOfKey(key));
                     }
 
                     Debug.Assert(elements != s_emptySentinel);
@@ -254,15 +254,10 @@ namespace Microsoft.CodeAnalysis.Collections
         /// </summary>
         private IDictionary<TKey, ImmutableArray<TElement>> EnsureFullyPopulated()
         {
-            IDictionary<TKey, ImmutableArray<TElement>>? fullyPopulatedMap = null;
-
             var currentMap = _map;
             while (IsNotFullyPopulatedMap(currentMap))
             {
-                if (fullyPopulatedMap == null)
-                {
-                    fullyPopulatedMap = CreateFullyPopulatedMap(currentMap);
-                }
+                IDictionary<TKey, ImmutableArray<TElement>> fullyPopulatedMap = CreateFullyPopulatedMap((ConcurrentDictionary<TKey, ImmutableArray<TElement>>?)currentMap);
 
                 var replacedMap = Interlocked.CompareExchange(ref _map, fullyPopulatedMap, currentMap);
                 if (replacedMap == currentMap)

--- a/src/Compilers/Core/Portable/Collections/CachingDictionary.cs
+++ b/src/Compilers/Core/Portable/Collections/CachingDictionary.cs
@@ -233,13 +233,7 @@ namespace Microsoft.CodeAnalysis.Collections
                 foreach (var key in allKeys)
                 {
                     // Copy non-empty values from the existing map
-                    ImmutableArray<TElement> elements;
-
-                    if (!existingMap.TryGetValue(key, out elements))
-                    {
-                        elements = existingMap.GetOrAdd(key, _getElementsOfKey(key));
-                    }
-
+                    ImmutableArray<TElement> elements = existingMap.GetOrAdd(key, _getElementsOfKey);
                     Debug.Assert(elements != s_emptySentinel);
                     fullyPopulatedMap.Add(key, elements);
                 }

